### PR TITLE
feat(GCS+gRPC): implement `SetBucketIamPolicy()`

### LINKED ISSUE #4160: Implement storage::GrpcClient::SetBucketIamPolicy

### PR DESCRIPTION
**Pull request by coryan**
Fixes #4160

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8431)

### CODE REVIEW

File: google/cloud/storage/internal/bucket_requests.h [hunk 1]
```diff
@@ -220,10 +220,12 @@ class SetBucketIamPolicyRequest
                                      google::cloud::IamPolicy const& policy);
 
   std::string const& bucket_name() const { return bucket_name_; }
+  google::cloud::IamPolicy const& policy() const { return policy_; }
```
Comment (dbolduc) [inline google/cloud/storage/internal/bucket_requests.h:223]: can we just say `IamPolicy` ?

File: google/cloud/storage/internal/bucket_requests.h [hunk 2]
```diff
@@ -220,10 +220,12 @@ class SetBucketIamPolicyRequest
...
   std::string const& json_payload() const { return json_payload_; }
 
  private:
   std::string bucket_name_;
+  google::cloud::IamPolicy policy_;
```
Comment (dbolduc) [inline google/cloud/storage/internal/bucket_requests.h:228]: same

File: google/cloud/storage/internal/grpc_bucket_request_parser.cc [hunk 3]
```diff
@@ -170,6 +170,50 @@ NativeIamPolicy GrpcBucketRequestParser::FromProto(
...
+  for (auto const& b : request.policy().bindings) {
+    google::iam::v1::Binding binding;
+    binding.set_role(b.first);
+    for (auto const& m : b.second) binding.add_members(m);
+    *policy.add_bindings() = std::move(binding);
+  }
+  return result;
+}
+
+#include "google/cloud/internal/diagnostics_pop.inc"
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc_bucket_request_parser.cc:215]: can we add the same `TODO(#5929)` here, to make super sure we won't forget to remove the line.

File: google/cloud/storage/internal/grpc_bucket_request_parser_test.cc [hunk 4]
```diff
@@ -298,6 +298,82 @@ TEST(GrpcBucketRequestParser, NativeIamPolicy) {
...
+
+  SetBucketIamPolicyRequest req("test-bucket", policy);
+  req.set_multiple_options(UserProject("test-user-project"),
+                           QuotaUser("test-quota-user"),
+                           UserIp("test-user-ip"));
+  auto const actual = GrpcBucketRequestParser::ToProto(req);
+  EXPECT_THAT(actual, IsProtoEqual(expected));
+}
+
+#include "google/cloud/internal/diagnostics_pop.inc"
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc_bucket_request_parser_test.cc:375]: same

File: google/cloud/storage/internal/grpc_client.cc [hunk 5]
```diff
@@ -242,14 +242,34 @@ StatusOr<NativeIamPolicy> GrpcClient::GetNativeBucketIamPolicy(
...
+  if (!response) return std::move(response).status();
+  IamBindings bindings;
+  for (auto const& b : response->bindings()) {
+    bindings.AddMembers(b.role(), std::set<std::string>(b.members().begin(),
+                                                        b.members().end()));
+  }
+  return IamPolicy{response->version(), std::move(bindings), response->etag()};
 }
 
+#include "google/cloud/internal/diagnostics_pop.inc"
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc_client.cc:263]: same

**Review by dbolduc — APPROVED** [review]

File: google/cloud/storage/internal/grpc_bucket_request_parser.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc_bucket_request_parser.cc:215]: We really need to remove the whole function, I changed the TODO comment to say that.

File: google/cloud/storage/internal/bucket_requests.h (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/bucket_requests.h]: Done.

File: google/cloud/storage/internal/bucket_requests.h (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/bucket_requests.h]: Done

File: google/cloud/storage/internal/grpc_bucket_request_parser_test.cc (hunk 4 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc_bucket_request_parser_test.cc:375]: Ditto

File: google/cloud/storage/internal/grpc_client.cc (hunk 5 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc_client.cc:263]: Ditto

**Pull request merged**